### ISSUE TITLE
Why EnumEntry is a class and not a trait?

### ISSUE THREAD

**Issue by joan38** [lloydmeta/enumeratum#116]
All in the title

Cheers

**Comment by lloydmeta** [comment]
Mainly to keep things simple. For something like `EnumEntry`, I didn't anticipate there being a real need for it to be a trait (e.g. not much of a use case for enums absolutely needing to inherit from another abstract class)

On the other hand, there are certain advantages: 

- Less bytecode: Traits result in more bytecode and some additional overhead related to the use of forwarder methods.
- Less recompiling for users: When a trait gains or loses a member, any classes that inherit from it must be recompiled, even if they have not changed. This is not the case with classes
- More efficient: Most Java runtimes make a virtual method invocation of a class member a faster operation than an interface method invocation. Traits get compiled to interfaces and therefore may pay a slight performance overhead.

http://stackoverflow.com/questions/35249901/scala-client-composition-with-traits-vs-implementing-an-abstract-class

**Comment by joan38** [comment]
I think all 3 points are not relevant anymore in 2.12

I find it really annoying to explicitly introduce limitation like that is a library that is supposed to be lightweight.
I mean this is not up to an enum lib to decide whether or not I should not inherit from another class.

**Comment by lloydmeta** [comment]
This library still supports 2.10 and 2.11, both of which still have a lot of users, so ignoring those versions is not an option.

> I think all 3 points are not relevant anymore in 2.12 with Java 8's defaults implementation.

I would love to see some file size and performance numbers backing that up. 

As far as I know 2.12.0 was released with released with the [same forwarding mechanism as 2.11](http://scala-lang.org/news/2.12.0#known-issues) for traits and I'm not aware of any fixes there in 2.12.1.

> I think this is not up to the enum lib to decide whether or not we should inherit from another class.

Note I didn't say should, just that I didn't anticipate much of a use case and went with the least power principle.

**Comment by joan38** [comment]
I see.
So are we staying with an abstract class? Or are you happy for me to do a PR for a trait?

**Comment by lloydmeta** [comment]
Until I see more convincing arguments in favour of `trait` than `abstract class`, I think it's best to stick with `abstract class` for now.

I'd encourage you to give the lib a try regardless though :) I've not felt limited by the definition of `EntryEnum` as a `trait` yet.

**Comment by joan38** [comment]
We are using it already. But it seems we will have to stop using it as this issue is a blocker.

I find it really sad that we get limited on this just because of peanut performance optimisation.

Also using a whole class is conceptually wrong to me, an enum entry is a trait or character of a class.

**Comment by lloydmeta** [comment]
I'm not very firm about on what you or I think is conceptually right or wrong, because there are arguments for both sides here.

Regarding performance, file-size and effect compilation, some real evidence would go a lot further in convincing me than telling me it's peanuts.

I also forgot that first point in the link I referenced earlier

> Traits complicate Java compatibility. If you have a trait with a companion object, calling methods on the companion object from Java requires bizarre MyType$.MODULE$.myMethod syntax. This isn't the case for abstract classes with companion objects, which are implemented on the JVM as a single class with static and instance methods. Implementing a Scala trait with concrete methods in Java is even more unpleasant.

**Comment by lloydmeta** [comment]
As far as performance is concerned I ran a few benchmarks that call a simple accessor on an `EnumEntry` and got a statistically significant but small increase of ~0.2-0.3 ns when using a `trait` instead of `abstract class`. Not a big deal in my books :)

I'll leave this issue open for a bit to hear more from you and maybe others before pulling the trigger.

-----
Results
```
Abstract class
    ----------------
    1st
    [info] Benchmark                         Mode  Cnt  Score   Error  Units
    [info] EnumBenchmarks.entryNameStacked   avgt   30  3.150 ± 0.023  ns/op
    [info] EnumBenchmarks.entryNameStandard  avgt   30  3.121 ± 0.029  ns/op
    
    2nd
    [info] Benchmark                         Mode  Cnt  Score   Error  Units
    [info] EnumBenchmarks.entryNameStacked   avgt   30  3.173 ± 0.037  ns/op
    [info] EnumBenchmarks.entryNameStandard  avgt   30  3.123 ± 0.024  ns/op
    
    3rd
    [info] Benchmark                         Mode  Cnt  Score   Error  Units
    [info] EnumBenchmarks.entryNameStacked   avgt   30  3.160 ± 0.037  ns/op
    [info] EnumBenchmarks.entryNameStandard  avgt   30  3.112 ± 0.037  ns/op
    
    Trait
    -----
    1st
    [info] Benchmark                         Mode  Cnt  Score   Error  Units
    [info] EnumBenchmarks.entryNameStacked   avgt   30  3.458 ± 0.091  ns/op
    [info] EnumBenchmarks.entryNameStandard  avgt   30  3.428 ± 0.054  ns/op
    
    2nd
    [info] Benchmark                         Mode  Cnt  Score   Error  Units
    [info] EnumBenchmarks.entryNameStacked   avgt   30  3.311 ± 0.056  ns/op
    [info] EnumBenchmarks.entryNameStandard  avgt   30  3.267 ± 0.039  ns/op
    
    3rd
    [info] Benchmark                         Mode  Cnt  Score   Error  Units
    [info] EnumBenchmarks.entryNameStacked   avgt   30  3.345 ± 0.047  ns/op
    [info] EnumBenchmarks.entryNameStandard  avgt   30  3.539 ± 0.093  ns/op
```

**Comment by joan38** [comment]
Nice benchmark @lloydmeta 
Let's see any other point of view on this.

Thanks

**Comment by lloydmeta** [comment]
No problem. I don't have a firm opinion either way, just want to make sure decisions are made rationally.

If you can share a real world use case (e.g. from drivetribe.com) that really won't work without `EnumEntry` being a `trait`, that would also help a lot.

**Comment by jvican** [comment]
> I think all 3 points are not relevant anymore in 2.12 with Java 8's defaults implementation.
  
Actually, point 1 and 2 still hold with 2.12. The first one is still true, forwarder methods are still created (but in another way) in 2.12. And the second one is up to Zinc, not the Scala version.
  
However, if this is really a blocker, I would also consider using a trait instead of an abstract class/class. The performance penalty seems minimal, though the issue with incremental compilation still holds. However, I would say that enumerations are not that prone to change in a codebase (they describe a data model that usually stays the same). Though I think the initial design decisions are correct and I also prefer classes over traits, I would also suggest making it a trait :wink:.

**Comment by lloydmeta** [comment]
I ran a small Twitter poll and although the number of responses was low, it was clear that there is support for making `EnumEntry` a trait, so I'll merge #118

**Comment by lloydmeta** [comment]
Released v1.5.8 just now which makes `EnumEntry` a trait. Thanks ! :)

**Comment by taylorwood** [comment]
This seems to be a breaking change if I have a transitive dependency on enumeratum 1.5.7 or lower.

1. I depend on library _foo_ which depends on enumeratum 1.5.7, and I want to depend on a later version of enumeratum
1. I exclude enumeratum when taking _foo_ dependency
1. I explicitly depend on a later version of enumeratum

Then I get the following error:
```
java.lang.IncompatibleClassChangeError: class path.to.MyType has interface enumeratum.EnumEntry as super class
```

Is there any way around this besides recompiling library _foo_ with a later version of enumeratum?
